dive_core/perf_metrics_data.cpp

```diff
@@ -223,18 +235,269 @@ PerfMetricsData::PerfMetricsData(std::vector<std::string>          metric_names,
+                             std::unordered_map<uint64_t, uint64_t>& out_mapping);
+
+    static bool IsSimilar(size_t n, const PerfMetricsRecord* a, const PerfMetricsRecord* b);
+
+    std::vector<DrawRef>                      m_draws;
+    std::unordered_map<DrawRef, PatternIndex> m_draw_to_index;
+
+    size_t                    m_pattern_size;
+    std::vector<PatternIndex> m_record_to_index;
+};
+
+std::optional<PerfMetricsDataProvider::Correlator::PatternIndex>
+PerfMetricsDataProvider::Correlator::MatchOf(size_t index) const
+{
+
+    if (index >= m_record_to_index.size())
+    {
+        return std::nullopt;
+    }
+    PatternIndex pattern_index = m_record_to_index[index];
+    if (pattern_index >= m_pattern_size)
+    {
+        return std::nullopt;
+    }
+    return pattern_index;
+}
+
+void PerfMetricsDataProvider::Correlator::ExtractDraws(
+const CommandHierarchy&                 command_hierarchy,
+std::vector<uint64_t>&                  out_draws,
+std::unordered_map<uint64_t, uint64_t>& out_mapping)
+{
+    std::unordered_map<uint64_t, uint64_t> draw_to_index;
+
+    std::vector<uint64_t> actual_draws;
+    std::vector<uint64_t> alias_draws;
+
+    std::vector<uint64_t>* draws = &actual_draws;
+
+    auto dedupe = [&]() {
+        draws = &actual_draws;
+        if (alias_draws.size() <= actual_draws.size())
+        {
+            size_t base = actual_draws.size() - alias_draws.size();
+            for (size_t i = 0; i < alias_draws.size(); ++i)
+            {
+                draw_to_index[alias_draws[i]] = base + i;
+            }
+        }
+        alias_draws.clear();
+    };
+
+    auto& alias_marker = command_hierarchy.GetFilterExcludeIndices(
+    Dive::CommandHierarchy::kBinningPassOnly);
+
+    for (size_t i = 0; i < command_hierarchy.size(); ++i)
+    {
+        auto node_type = command_hierarchy.GetNodeType(i);
+        if (node_type == Dive::NodeType::kSubmitNode)
+        {
+            dedupe();
+        }
+        if (node_type == Dive::NodeType::kRenderMarkerNode)
+        {
+            dedupe();
+            if (alias_marker.find(i) != alias_marker.end())
+            {
+                draws = &alias_draws;
+            }
+        }
+        if (!Dive::IsDrawDispatchNode(node_type))
+        {
+            continue;
+        }
+        draws->push_back(i);
+    }
+    dedupe();
+    for (size_t i = 0; i < actual_draws.size(); ++i)
+    {
+        draw_to_index[actual_draws[i]] = i;
+    }
+    out_draws = std::move(actual_draws);
+    out_mapping = std::move(draw_to_index);
+}
+
+void PerfMetricsDataProvider::Correlator::Reset()
+{
+    m_draws.clear();
+    m_draw_to_index.clear();
+
+    m_pattern_size = 0;
+    m_record_to_index.clear();
+}
+
+void PerfMetricsDataProvider::Correlator::AnalyzeCommands(const CommandHierarchy& command_hierarchy)
+{
+    ExtractDraws(command_hierarchy, m_draws, m_draw_to_index);
+}
+
+bool PerfMetricsDataProvider::Correlator::IsSimilar(size_t                   n,
```

Comment:
i feel that the name IsSimilar is a bit confusing, maybe something like `AreDrawSignaturesMatching` ?

Reply:
Updated.